src/subdomains/core/fee/fee.service.ts

```diff
@@ -0,0 +1,230 @@
+          if (userDiscount < fee.value) userDiscount = fee.value;
+          break;
+      }
+    }
+
+    if (!userBaseFee) throw new InternalServerErrorException('Base Fee is missing');
+    if (userBaseFee - userDiscount < 0) {
+      this.logger.warn(`UserDiscount higher userBaseFee! UserDataId: ${userDataId}`);
+      userDiscount = 0;
+    }
+
+    return userCustomFee ?? userBaseFee - userDiscount;
+  }
+
+  private async getValidUserFees(request: UserFeeRequest): Promise<Fee[]> {
+    const userFees: Fee[] = [];
+    const accountType = request.userData.accountType;
+
+    userFees.push(
+      ...(await this.getBaseFees({ ...request, accountType })),
+      ...(await this.getFreeDiscounts({ ...request, accountType })),
+    );
+
+    const discountCodes = request.userData.discounts?.split(';') ?? [];
+
+    for (const feeId of discountCodes) {
+      const fee = await this.feeRepo.findOneBy({ id: +feeId });
+
+      if (this.isExpiredFee(fee, { ...request, accountType })) {
+        await this.userDataService.removeDiscountCode(request.userData, fee.id.toString());
+        continue;
+      }
+
+      if (!this.isValidFee(fee, { ...request, accountType })) continue;
+
+      userFees.push(fee);
+    }
+
+    return userFees.sort((a, b) => {
+      const TypeOrder: { [b in FeeType]: number } = {
+        [FeeType.CUSTOM]: 1,
+        [FeeType.BASE]: 2,
+        [FeeType.DISCOUNT]: 3,
+      };
+      return TypeOrder[a.type] - TypeOrder[b.type];
+    });
+  }
+
+  private async getBaseFees(request: FeeRequest): Promise<Fee[]> {
+    const baseFees = await this.feeRepo.findBy([
+      { type: FeeType.BASE, accountType: request.accountType },
+      { type: FeeType.BASE, accountType: IsNull() },
+    ]);
+
+    return baseFees.filter((baseFee) => this.isValidFee(baseFee, { ...request, accountType: request.accountType }));
+  }
+
+  private async getFreeDiscounts(request: FeeRequest): Promise<Fee[]> {
+    const discounts = await this.feeRepo.findBy([
+      {
+        type: FeeType.DISCOUNT,
+        accountType: request.accountType,
+        discountCode: IsNull(),
+      },
+      {
+        type: FeeType.DISCOUNT,
+        accountType: IsNull(),
+        discountCode: IsNull(),
+      },
+    ]);
+
+    return discounts.filter((discount) => this.isValidFee(discount, { ...request, accountType: request.accountType }));
+  }
+
+  private isValidFee(fee: Fee, request: FeeRequest): boolean {
+    return !(
+      this.isExpiredFee(fee, request) ||
+      (fee.direction && fee.direction !== request.direction) ||
+      (fee.assets?.length > 0 && request.asset && !fee.assets.split(';').includes(request.asset?.id.toString())) ||
+      (fee.maxTxVolume && fee.maxTxVolume < request.txVolume)
+    );
+  }
+
+  private isExpiredFee(fee: Fee, request: FeeRequest): boolean {
+    return (
+      !fee ||
+      (fee.expiryDate && fee.expiryDate < new Date()) ||
+      (fee.accountType && fee.accountType !== request.accountType)
+    );
+  }
```

Comment:
Can we just combine these two methods and rename to `isMatchingFee`? Or remove the accountType check on `isExpiredFee`?